pkg/controller/daemon/daemon_controller.go

```diff
@@ -80,6 +80,17 @@ const (
 // controllerKind contains the schema.GroupVersionKind for this controller type.
 var controllerKind = apps.SchemeGroupVersion.WithKind("DaemonSet")
 
+// NodeUpdateItem represents a node update event that may affect DaemonSets.
+// It contains the node name and previous node state (labels and taints) to determine
+// if DaemonSet pods need to be added or removed from the node after the update.
+// Note: When adding new fields to NodeUpdateItem, ensure to update both the struct definition
+// and the enqueueNodeUpdate function to handle the new fields properly.
+type NodeUpdateItem struct {
```

Comment:
I'm not convinced it will work like that.

The problem is that underlying Queue does deduplication of the same items.
Here the item is the whole item, (together with labels & taints).

But if the node quickly changes more than once, you have no guarantee that a given node (just in two different states) will not be processed by 2 different workers, leading to races and potentially messing up the cluster state.
Imagine the case that:
- update1 should trigger DS pod creation
- update2 should trigger DS pod deletion
Now, if we process them in the opposite order (which might be the case due to those being handled by two different workers), we can end-up with pod being first deleted and only then created and we will leak the pod...

We need some better way of solving this.

Reply:
A simple implementation: when a node event is detected, all DaemonSets are directly enqueued. This approach resolves race conditions. However, every node change triggers the reprocessing of all DaemonSets, which may lead to unnecessary processing and low efficiency. That said, in most cases, the number of DaemonSets is relatively small, and node updates are infrequent, making this simple approach generally feasible while effectively avoiding complex race conditions.